components/api-resource-mgt/org.wso2.carbon.identity.api.resource.collection.mgt/src/main/java/org/wso2/carbon/identity/api/resource/collection/mgt/listener/ConsoleRoleListener.java

```diff
@@ -0,0 +1,256 @@
+    public void postGetRole(Role role, String roleId, String tenantDomain) throws IdentityRoleManagementException {
+
+
+        if (isConsoleRole(roleId, tenantDomain)) {
+            // Get updated console role permissions with newly added read and write scopes from API resource collection.
+            List<Permission> rolePermissions = getUpdatedPermissions(role.getPermissions(), tenantDomain);
+            role.setPermissions(rolePermissions);
+        }
+    }
+
+    @Override
+    public void postGetPermissionListOfRole(List<Permission> permissionListOfRole, String roleId, String tenantDomain)
+            throws IdentityRoleManagementException {
+
+        if (isConsoleRole(roleId, tenantDomain)) {
+            List<Permission> rolePermissions = getUpdatedPermissions(permissionListOfRole, tenantDomain);
+            permissionListOfRole.retainAll(rolePermissions);
+        }
+    }
+
+    @Override
+    public void postGetPermissionListOfRoles(List<String> permissions, List<String> roleIds, String tenantDomain)
+            throws IdentityRoleManagementException {
+
+        boolean isConsoleRole = false;
+        for (String roleId : roleIds) {
+            if (isConsoleRole(roleId, tenantDomain)) {
+                isConsoleRole = true;
+                break;
+            }
+        }
+        if (isConsoleRole) {
+            List<Permission> resolvedRolePermissions = new ArrayList<>();
+            List<Permission> systemPermissions = getSystemPermission(tenantDomain);
+            permissions.forEach(permission -> {Optional<Permission> newPermission = systemPermissions.stream()
+                        .filter(permission1 -> permission1.getName().equals(permission)).findFirst();
+                newPermission.ifPresent(resolvedRolePermissions::add);
+            });
+            List<Permission> rolePermissions = getUpdatedPermissions(resolvedRolePermissions, tenantDomain);
+            permissions.retainAll(rolePermissions.stream().map(Permission::getName).collect(Collectors.toList()));
+        }
+    }
+
+    @Override
+    public void preUpdatePermissionsForRole(String roleId, List<Permission> addedPermissions,
+                                            List<Permission> deletedPermissions, String audience, String audienceId,
+                                            String tenantDomain) throws IdentityRoleManagementException {
+
+        if (APPLICATION.equals(audience)) {
+            List<Permission> consoleFeaturePermissions = getConsoleFeaturePermissions(addedPermissions);
+            if (ArrayUtils.isNotEmpty(consoleFeaturePermissions.toArray())) {
+                // If console features are added to the role, then we need to we only need to persist the console
+                // permissions.
+                addedPermissions.retainAll(consoleFeaturePermissions);
+            }
+        }
+    }
+
+    private List<Permission> getUpdatedPermissions(List<Permission> rolePermissions, String tenantDomain)
```

Comment:
A Suggession to organize the code.
------------

Filter permissions list starts with
Iterate over collection:     console:app, console:idp,.....
     if console:app:edit exist in permission list.
           Add all respective scopes. 
           Continue.
      if console:app:view exist in permission list.
           Add respective view scopes. 
           Continue.
      // backward compatible logic.
      x = Fetch previous console:app edit related scopes set.
      if x contains in permission list
            Add all respective scopes. 
            Continue.
     
       Add respective view scopes. 
       Continue.

Reply:
+1